compiler/rustc_hir_analysis/src/check/errs.rs

```diff
@@ -2,79 +2,145 @@ use rustc_hir as hir;
-    if let hir::ExprKind::AddrOf(borrow_kind, m, expr) = expr.kind
-        && matches!(borrow_kind, hir::BorrowKind::Ref)
-        && let Some(var) = path_if_static_mut(tcx, expr)
-    {
-        handle_static_mut_ref(tcx, span, var, span.edition().at_least_rust_2024(), m, hir_id);
+pub fn maybe_expr_static_mut(tcx: TyCtxt<'_>, expr: &hir::Expr<'_>) {
+    let err_span = expr.span;
+    let lint_level_hir_id = expr.hir_id;
+    match expr.kind {
+        hir::ExprKind::AddrOf(borrow_kind, m, ex)
+            if matches!(borrow_kind, hir::BorrowKind::Ref)
+                && let Some((var, err_span)) = path_if_static_mut(tcx, ex, err_span) =>
+        {
+            handle_static_mut_ref(
+                tcx,
+                err_span,
+                var,
+                err_span.edition().at_least_rust_2024(),
+                m,
+                lint_level_hir_id,
+                !expr.span.from_expansion(),
+            );
+        }
+        hir::ExprKind::Index(expr, _, _)
+            if let Some((var, err_span)) = path_if_static_mut(tcx, expr, err_span) =>
+        {
+            handle_static_mut_ref(
+                tcx,
+                err_span,
+                var,
+                err_span.edition().at_least_rust_2024(),
+                Mutability::Not,
+                lint_level_hir_id,
+                false,
+            );
+        }
+        _ => {}
     }
 }
 
 /// Check for shared or mutable references of `static mut` inside statement
-pub fn maybe_stmt_static_mut(tcx: TyCtxt<'_>, stmt: hir::Stmt<'_>) {
+pub fn maybe_stmt_static_mut(tcx: TyCtxt<'_>, stmt: &hir::Stmt<'_>) {
     if let hir::StmtKind::Let(loc) = stmt.kind
         && let hir::PatKind::Binding(ba, _, _, _) = loc.pat.kind
         && let hir::ByRef::Yes(rmutbl) = ba.0
         && let Some(init) = loc.init
-        && let Some(var) = path_if_static_mut(tcx, init)
+        && let Some((var, err_span)) = path_if_static_mut(tcx, init, init.span)
     {
         handle_static_mut_ref(
             tcx,
-            init.span,
+            err_span,
             var,
             loc.span.edition().at_least_rust_2024(),
             rmutbl,
             stmt.hir_id,
+            false,
+        );
+    }
+}
+
+/// Check if method call takes shared or mutable references of `static mut`
+pub fn maybe_method_static_mut(tcx: TyCtxt<'_>, expr: &hir::Expr<'_>) {
+    if let hir::ExprKind::MethodCall(_, e, _, _) = expr.kind
+        && let Some((var, err_span)) = path_if_static_mut(tcx, e, expr.span)
+        && let typeck = tcx.typeck(expr.hir_id.owner)
+        && let Some(method_def_id) = typeck.type_dependent_def_id(expr.hir_id)
+        && let inputs = tcx.fn_sig(method_def_id).skip_binder().inputs().skip_binder()
+        && !inputs.is_empty()
+        && inputs[0].is_ref()
+    {
+        handle_static_mut_ref(
+            tcx,
+            err_span,
+            var,
+            err_span.edition().at_least_rust_2024(),
+            Mutability::Not,
```

Comment:
@obeis, can you make the mutability parameter here depend on whether `inputs[0]` is `&` or `&mut`?